File: rs/tests/divergence_test.rs
// The recorded divergences, pinned.
//
// Every entry in ../../DIVERGENCE.md with a RUST column that a shared
// fixture cannot express has a test here. They fail on REPAIR as loudly
// as on regression: a divergence that quietly goes away should be
// deleted from the register in the same change, not left behind as a
// claim nobody checks.
//
// Three are pinned elsewhere, beside the code they belong to: UNDEFINED
// against null, in `undefined_test.rs`; the unpaired surrogate table of
// "UTF-16 escapes in a double quoted scalar", in `js_semantics_test.rs`;
// and that entry's second table, the window cut through an astral
// character, in `escape_window_test.rs`.
//
// One entry has no Rust column to pin. "A tab-only tail at the end of
// the source" is a GO divergence: this port gives the canonical answer,
// the refusal included, so the row that can fail lives in
// `go/divergence_test.go`.

mod common;

use common::plain;
use serde_json::json as j;

/// **Nesting past 127 containers is refused**, with the engine's
/// `cancel` code, where TypeScript and Go have no limit.
///
/// A parse guard this plugin installs over jsonic's, at jsonic's number,
/// counting jsonic's containers and YAML's own block collections. The
/// parse loop is iterative, but displaying, converting or dropping a
/// value walks the tree with the call stack, so an unbounded document
/// ends the process rather than returning an error. See jsonic's own
/// register for the measurements.
#[test]
fn nesting_past_the_depth_limit_is_refused() {
    let parser = tabnas_yaml::make();
    let within = format!("{}{}", "[".repeat(127), "]".repeat(127));
    parser
        .parse(&within)
        .expect("127 containers is inside the limit");
    let past = format!("{}{}", "[".repeat(128), "]".repeat(128));
    let error = parser.parse(&past).expect_err("128 containers is past it");
    assert_eq!(error.code, "cancel");
    // A compact block sequence nests through YAML's own rules.
    parser
        .parse(&format!("{}x\n", "- ".repeat(127)))
        .expect("127 compact sequences is inside the limit");
    let error = parser
        .parse(&format!("{}x\n", "- ".repeat(128)))
        .expect_err("128 compact sequences is past it");
    assert_eq!(error.code, "cancel");
}

/// **A column counts Unicode scalars, not UTF-16 code units.** An astral
/// character is one column here and in Go, and two in TypeScript.
///
/// Inherited from the engine: see parser/DIVERGENCE.md, "Column
/// positions for astral characters". Measured:
///
///  input            TypeScript   Go   Rust
///  `{a: <emoji>, ]`    10         9     9
#[test]
fn an_astral_character_is_one_column() {
    let error = tabnas_yaml::make()
        .parse("{a: \u{1F600}, ]")
        .expect_err("the bracket does not close the brace");
    assert_eq!(
        error.col, 9,
        "TypeScript reports 10 because it counts UTF-16 code units"
    );
    // The control: the same document with an ASCII value agrees exactly.
    let ascii = tabnas_yaml::make()
        .parse("{a: x, ]")
        .expect_err("the bracket does not close the brace");
    assert_eq!(ascii.col, 9);
}

/// **A tag whose value is a `%TAG` directive line: REPAIRED, kept as a
/// parity assertion.**
///
/// This was the one VALUE difference the matcher-chain entry recorded,
/// and it was not the matcher chain at all. The typed tag's unquoted
/// value ends at a colon only before a literal SPACE, a line end or the
/// end of the source, and this port ended it before a TAB as well, so
/// the colon reached the grammar as a key separator and the first
/// document became a mapping. `../DIVERGENCE.md` no longer carries the
/// row, and `test/spec/directives.tsv` carries the input instead.
///
/// Measured 2026-09-21, all three runtimes:
/// `!!int%TAG !! x:\t\n... ... %YAML 1.20o17` is
/// `["!! x:", "... %YAML 1.20o17"]` in TypeScript, in Go and here.
#[test]
fn a_tag_before_a_directive_line_resolves_the_canonical_way() {
    let value = tabnas_yaml::make()
        .parse("!!int%TAG !! x:\t\n... ... %YAML 1.20o17 ")
        .expect("the document parses in every runtime");
    assert_eq!(
        plain(&value),
        j!(["!! x:", "... %YAML 1.20o17"]),
        "the first document is the string \"!! x:\", not a mapping"
    );
}

/// **A refusal both runtimes make can be reported at a different place.**
///
/// Same cause as above: this port refuses at the character its own
/// matcher cannot claim, and the canonical one carries on to a later
/// one. Both refuse, with the same code; only the position differs, and
/// no shared fixture pins a position.
///
/// Measured:
///
///  input                            TypeScript   Rust
///  `[a, b\n{a: 1\n? k\n- x\n`         5:1         3:1
///  `---\n{a: 1\na: 1\n  t2\n`         5:1         4:3
#[test]
fn a_shared_refusal_can_be_reported_at_a_different_place() {
    let parser = tabnas_yaml::make();
    for (src, row, col) in [
        ("[a, b\n{a: 1\n? k\n- x\n", 3, 1),
        ("---\n{a: 1\na: 1\n  t2\n", 4, 3),
    ] {
        let error = parser.parse(src).expect_err("both runtimes refuse this");
        assert_eq!(error.code, "unexpected", "{src:?}");
        assert_eq!((error.row, error.col), (row, col), "{src:?}");
    }
}

/// **A blank line of tabs at the end of the source can cost a token.**
///
/// The same end-of-source guard, reached a different way. A line holding
/// nothing but blanks, one of them a TAB, is skipped by both matchers:
/// YAML counts it blank and the engine refuses a bare tab. The canonical
/// then moves its point and returns no token, and the engine's own
/// matchers take over at the end of the source. This port returns
/// `Act::refuse`, and the explicit key never gets the null it owes.
///
/// Making the guard refuse only when nothing was consumed was measured
/// and is NOT the repair: it lets `a: !é` reach the engine with an
/// `internal` error, which `untrusted_test.rs` catches.
///
/// Measured 2026-09-21:
///
///  input               TypeScript     Go             Rust
///  `? k\n: <TAB>\n`    `{"k":null}`   `{"k":null}`   `ERROR:unexpected`
///  `? k\n: <TAB>`      `{"k":null}`   `{"k":null}`   `ERROR:unexpected`
///  `? k\n:   <TAB>\n`  `{"k":null}`   `{"k":null}`   `ERROR:unexpected`
///  `? k\n:  \n`        `{"k":null}`   `{"k":null}`   `{"k":null}`
///  `? k\n: <TAB>v\n`   `{"k":"v"}`    `{"k":"v"}`    `{"k":"v"}`
#[test]
fn a_trailing_blank_line_of_tabs_costs_an_explicit_keys_value() {
    let parser = tabnas_yaml::make();
    for src in ["? k\n: \t\n", "? k\n: \t", "? k\n:   \t\n"] {
        let error = parser
            .parse(src)
            .expect_err("TypeScript and Go both accept this one");
        assert_eq!(error.code, "unexpected", "{src:?}");
    }

    // The controls: the same shapes without the tab, and the same tab
    // with a value after it, agree in all three runtimes. Without them
    // the rows above would also pass on a port that refused every
    // explicit key.
    assert_eq!(
        plain(&parser.parse("? k\n:  \n").expect("parses")),
        j!({"k": null})
    );
    assert_eq!(
        plain(&parser.parse("? k\n: \tv\n").expect("parses")),
        j!({"k": "v"})
    );
    assert_eq!(
        plain(&parser.parse("? k\n: \t\nz: 1\n").expect("parses")),
        j!({"k": {"z": 1}})
    );
}

/// **A flow collection or a quoted scalar at column 0 after a line
/// break is refused**, in every runtime.
///
/// Not a divergence: a shape that looks like one, kept here because it
/// is surprising and because a port that "fixed" it would diverge. The
/// canonical matcher skips the indent token before such a character, and
/// the engine's remaining matchers were chosen for the line break, so the
/// bracket is unclaimed. A quoted scalar is fine, because this plugin's
/// own matcher lexes it.
#[test]
fn a_flow_collection_at_column_zero_after_a_line_break_is_refused() {
    let parser = tabnas_yaml::make();
    for src in ["# c\n{a: 1}", "   \n[1, 2]", "\n{a: 1}"] {
        let error = parser.parse(src).expect_err("every runtime refuses this");
        assert_eq!(error.code, "unexpected", "{src:?}");
    }
    // The quoted forms do parse, which is what makes the above a
    // property of the matcher chain rather than of column 0.
    assert_eq!(plain(&parser.parse("   \n\"x\"").expect("parses")), j!("x"));
    // A flow collection on the FIRST line is fine, so the refusal is
    // about what the lexer call dispatched on and not about brackets.
    assert_eq!(
        plain(&parser.parse("{a: 1}").expect("parses")),
        j!({"a": 1})
    );
    // The position is the canonical one too, quirk included: the
    // canonical matcher sets the column to zero before the marker it
    // skipped, and this port carries that through.
    let error = parser.parse("# c\n{a: 1}").expect_err("refused");
    assert_eq!((error.row, error.col), (2, 0));
}

/// **An unterminated typed tag folds a split astral character** to the
/// replacement character, where TypeScript keeps half of one.
///
/// The canonical handler ends an unterminated `!!str "` at the end of the
/// source and takes `fwd.substring(valStart + 1, valEnd - 1)`, whose
/// indices count UTF-16 CODE UNITS: the value it keeps is everything up
/// to the last unit. An astral character is two of those units and one
/// Rust scalar, so dropping one unit leaves a LONE HIGH SURROGATE, which
/// a Rust string cannot hold; this port folds it exactly as it folds an
/// escape naming one. A character inside the Basic Multilingual Plane is
/// one unit and one scalar, and agrees exactly.
///
/// Same trade as "UTF-16 escapes in a double quoted scalar", and the same
/// owner: the string model. Measured, with `U+XXXX` for what the table is
/// about, because a lone surrogate has no UTF-8 spelling:
///
///  input                        TypeScript        Go                   Rust
///  `a: !!str "<emoji>`          U+D83D            three U+FFFD         U+FFFD
///  `a: !!str "a<emoji>`         `a` then U+D83D   `a`, three U+FFFD    `a` U+FFFD
///  `a: !!str "<U+4E2D><U+6587>` U+4E2D            U+4E2D, two U+FFFD   U+4E2D
///  `a: !!str "`                 `"`               refused, `internal`  `"`
#[test]
fn an_unterminated_typed_tag_folds_a_split_astral_character() {
    let parser = tabnas_yaml::make();
    for (src, want) in [
        ("a: !!str \"\u{1F600}", "\u{FFFD}"),
        ("a: !!str \"a\u{1F600}", "a\u{FFFD}"),
        ("a: !!str '\u{1F600}", "\u{FFFD}"),
    ] {
        let value = parser
            .parse(src)
            .unwrap_or_else(|error| panic!("{src:?}: {error}"));
        assert_eq!(
            plain(&value),
            j!({ "a": want }),
            "{src:?}: TypeScript keeps the lone high surrogate"
        );
    }

    // The control. A tail inside the Basic Multilingual Plane is one
    // UTF-16 unit and one Rust scalar, so the two runtimes agree, and so
    // does the reversed-index case a `substring` swaps.
    for (src, want) in [
        ("a: !!str \"\u{4e2d}\u{6587}", "\u{4e2d}"),
        ("a: !!str \"\u{e9}", ""),
        ("a: !!str \"ab", "a"),
        ("a: !!str \"", "\""),
        ("a: !!str '", "'"),
    ] {
        let value = parser
            .parse(src)
            .unwrap_or_else(|error| panic!("{src:?}: {error}"));
        assert_eq!(plain(&value), j!({ "a": want }), "{src:?}");
    }
}

File: go/divergence_test.go
// Copyright (c) 2026 Richard Rodger and other contributors, MIT License

package tabnasyaml

// The Go column of ../DIVERGENCE.md, made executable.
//
// Every entry in that file is a row this port answers differently from
// the canonical TypeScript, and a register that is only prose records a
// claim rather than a behaviour: it cannot fail when the port drifts,
// and it cannot fail when the port is REPAIRED, which is the moment the
// entry has to be deleted. The Rust half of each table is pinned in
// rs/tests, and the Go half was not pinned anywhere until this file.
//
// None of these rows can be a shared test/spec/*.tsv fixture. An
// expected cell is UTF-8 text and the TypeScript answers here are lone
// UTF-16 surrogates, which have no UTF-8 spelling; the last test is a
// case the canonical REFUSES and this port accepts, and a fixture row
// carries one answer for all three runtimes.
//
// Measured on 2026-09-22: the TypeScript column by running
// ts/src/yaml.ts under Node 22, the Go column by Parse below, the Rust
// column by tabnas_yaml::parse.
//
// A test here failing because the port now AGREES with the canonical is
// the good direction, and the response is to delete the row, its entry
// in ../DIVERGENCE.md and the matching Rust pin together.

import (
	"encoding/json"
	"testing"
)

// parseJSON parses src and returns the result as JSON text.
//
// JSON text rather than a Go value, because these rows are about which
// CHARACTERS end up in a string, and a %#v of a map hides them behind
// escape spellings that differ between the two ports.
func parseJSON(t *testing.T, src string) string {
	t.Helper()
	value, err := Parse(src)
	if err != nil {
		t.Fatalf("parse %q: %v", src, err)
	}
	encoded, merr := json.Marshal(value)
	if merr != nil {
		t.Fatalf("marshal %q: %v", src, merr)
	}
	return string(encoded)
}

// TestAnUnpairedSurrogateEscapeFolds pins the first table of "UTF-16
// escapes in a double quoted scalar".
//
// A `\u` escape is String.fromCharCode, which appends a UTF-16 CODE
// UNIT, and a JavaScript string holds an unpaired one. A Go string holds
// Unicode scalars, so this port folds the half to the replacement
// character. The controls are the paired rows, which reach the canonical
// astral character and are shared fixtures besides: without them "every
// escape folds" would also satisfy this test.
func TestAnUnpairedSurrogateEscapeFolds(t *testing.T) {
	for _, row := range []struct {
		src  string
		want string
		ts   string // for the reader: the canonical answer
	}{
		{`a: "\uD83D"`, "{\"a\":\"\ufffd\"}", "a lone high surrogate"},
		{`a: "\uDE00"`, "{\"a\":\"\ufffd\"}", "a lone low surrogate"},
		{`a: "\U0000D83D"`, "{\"a\":\"\ufffd\"}", "a lone high surrogate"},

		// The controls: a pair is one astral character in both ports.
		{`a: "\uD83D\uDE00"`, "{\"a\":\"\U0001F600\"}", "U+1F600"},
		{`a: "\U0000D83D\uDE00"`, "{\"a\":\"\U0001F600\"}", "U+1F600"},
	} {
		if got := parseJSON(t, row.src); got != row.want {
			t.Errorf("%q = %s, want %s (TypeScript gives %s). A repair here "+
				"retires the first table of \"UTF-16 escapes in a double "+
				"quoted scalar\" in ../DIVERGENCE.md", row.src, got, row.want, row.ts)
		}
	}
}

// TestASplitAstralEscapeWindowFoldsTheHalfItLeaves pins the second table
// of the same entry.
//
// The window a `\x`, `\u` or `\U` escape reads is a fixed width counted
// in UTF-16 CODE UNITS, so it can end INSIDE an astral character. The
// canonical keeps the high surrogate inside the window, where it is no
// hexadecimal digit and ends parseInt's prefix, and the scan then
// appends the low surrogate as a character of its own. This port folds
// that half, as it folds every unpaired surrogate.
//
// The controls are the last two rows, and they are the reason this is a
// fold and not a window error: where the escape names a high surrogate
// itself the cut half COMPLETES the pair and the whole astral character
// survives, and a character inside the Basic Multilingual Plane is one
// unit and cannot be cut at all. Both would break if the window or the
// cursor drifted by a unit.
func TestASplitAstralEscapeWindowFoldsTheHalfItLeaves(t *testing.T) {
	for _, row := range []struct {
		src  string
		want string
		ts   string
	}{
		{"a: \"\\xA\U0001F600Z\"", "{\"a\":\"\\n\ufffdZ\"}", "U+000A U+DE00 Z"},
		{"a: \"\\u004\U0001F600Z\"", "{\"a\":\"\\u0004\ufffdZ\"}", "U+0004 U+DE00 Z"},
		{"a: \"\\U0000004\U0001F600Z\"", "{\"a\":\"\\u0004\ufffdZ\"}", "U+0004 U+DE00 Z"},

		// The controls, which agree with the canonical.
		{"a: \"\\U000D83D\U0001F600Z\"", "{\"a\":\"\U0001F600Z\"}", "U+1F600 Z"},
		{"a: \"\\xA\u4E2DZ\"", "{\"a\":\"\\nZ\"}", "U+000A Z"},
	} {
		if got := parseJSON(t, row.src); got != row.want {
			t.Errorf("%q = %s, want %s (TypeScript gives %s). A control row "+
				"failing means the WINDOW or the cursor moved, not the fold",
				row.src, got, row.want, row.ts)
		}
	}
}

// TestAnUnterminatedTypedTagFoldsASplitAstralCharacter pins the entry of
// that name.
//
// An unterminated quoted value after a typed tag runs to the end of the
// source, and the canonical then takes everything up to the last UTF-16
// code unit. An astral character is two of those units and one Go rune,
// so dropping one unit leaves a lone high surrogate, which this port
// folds.
//
// The controls are the last two rows: a Basic Multilingual Plane
// character is one unit, so the drop takes the whole of it, and an
// empty value leaves the substring's indices reversed, where the
// canonical swaps them and yields the quote itself.
func TestAnUnterminatedTypedTagFoldsASplitAstralCharacter(t *testing.T) {
	for _, row := range []struct {
		src  string
		want string
		ts   string
	}{
		{"a: !!str \"\U0001F600", "{\"a\":\"\ufffd\"}", "U+D83D"},
		{"a: !!str \"a\U0001F600", "{\"a\":\"a\ufffd\"}", "a then U+D83D"},
		{"a: !!str '\U0001F600", "{\"a\":\"\ufffd\"}", "U+D83D"},

		// The controls, which agree with the canonical.
		{"a: !!str \"\u4E2D\u6587", "{\"a\":\"\u4E2D\"}", "U+4E2D"},
		{"a: !!str \"", `{"a":"\""}`, `"`},
	} {
		if got := parseJSON(t, row.src); got != row.want {
			t.Errorf("%q = %s, want %s (TypeScript gives %s)",
				row.src, got, row.want, row.ts)
		}
	}
}

// TestATabOnlyTailAtTheEndOfTheSourceIsAccepted pins "A tab-only tail at
// the end of the source".
//
// The canonical's keyword and number branches advance by the TRIMMED
// text, so a tab after such a value stays in the source; the matcher's
// blank-line skip then consumes it, and the document is refused anyway,
// at the end of the source. Which step refuses it is not isolated:
// holding the row and the column still over that skip changes no row of
// the table. This port has no such branch, the tail reaches jsonic's own
// lexer, which reads a tab as a space, and the document ends.
//
// The controls are the rows both runtimes accept: the same tab with a
// newline after it, which is a blank line everywhere, a trailing SPACE,
// which enters no such branch, and a plain scalar, whose own handler has
// consumed the tab before the branch is reached.
func TestATabOnlyTailAtTheEndOfTheSourceIsAccepted(t *testing.T) {
	for _, row := range []struct {
		src  string
		want string
	}{
		{"a: 1\t", `{"a":1}`},
		{"a: true\t", `{"a":true}`},
		{"a: null\t", `{"a":null}`},
		{"a: [1]\t", `{"a":[1]}`},
		{"- 1\t", `[1]`},
		{"a: 1 \t", `{"a":1}`},

		// The controls, which the canonical accepts too.
		{"a: 1\t\nb: 2", `{"a":1,"b":2}`},
		{"a: 1 ", `{"a":1}`},
		{"a: x\t", `{"a":"x"}`},
	} {
		if got := parseJSON(t, row.src); got != row.want {
			t.Errorf("%q = %s, want %s", row.src, got, row.want)
		}
	}

	// The divergence itself: the canonical REFUSES the six rows above,
	// and this port must keep answering them until the canonical is
	// repaired. A control row failing is a plain regression; a first-group
	// row failing means this port now refuses them too, and the entry in
	// ../DIVERGENCE.md is the thing to delete.
	if _, err := Parse("a: 1\t"); err != nil {
		t.Fatalf("`a: 1<TAB>` is refused here now (%v), which is the "+
			"canonical answer: delete \"A tab-only tail at the end of the "+
			"source\" from ../DIVERGENCE.md and this test with it", err)
	}
}
